src/manifolds/Segre.jl

```diff
@@ -0,0 +1,473 @@
+
+    @assert(length(X_) == 0)
+    check_vector(M, p, v)
+
+    return v
+end
+
+@doc raw"""
+    function inner(M::Segre{ℝ, V}, p, u, v,)
+
+Inner product between two tangent vectors ``u = (\nu, u_1, \dots, u_d)`` and ``v = (\xi, v_1, \dots, v_d)`` at ``p \doteq (\lambda, x_1, \dots, x_d)``. This inner product is obtained by embedding the Segre manifold in the space of tensors equipped with the Euclidean metric:
+````math
+    \langle u, v \rangle_{p} = \nu \xi + \lambda^2 (\langle u_1, v_1 \rangle_{x_1} + \dots + \langle u_d, v_d \rangle_{x_d}),
+````
+where ``\nu``, ``\xi \in T_{\lambda} ℝ^{+} = ℝ`` and ``u_i``, ``v_i \in T_{x_i} S^{n_i - 1} \subset ℝ^{n_i}``.
+"""
+function inner(M::Segre{ℝ,V}, p, u, v) where {V}
+    return u[1][1] * v[1][1] + p[1][1]^2 * dot(u[2:end], v[2:end])
+end
+
+@doc raw"""
+    function norm(M::Segre{𝔽, V}, p, v)
+
+Norm of tangent vector ``v`` at ``p``.
+"""
+function norm(M::Segre{𝔽,V}, p, v) where {𝔽,V}
+    return sqrt(inner(M, p, v, v))
+end
+
+"""
+    function rand(M::Segre{ℝ, V}; vector_at=nothing)
+"""
+function rand(M::Segre{ℝ,V}; vector_at=nothing) where {V}
+    if isnothing(vector_at)
+        lambda = abs.(rand(Euclidean(1)))
+        xs = [rand(Sphere(n - 1)) for n in V]
+        return [lambda, xs...]
+    else
+        @assert(is_point(M, vector_at))
+
+        lambdadot = rand(Euclidean(1); vector_at=vector_at[1])
+        xdots = [rand(Sphere(n - 1); vector_at=vector_at[i + 1]) for (i, n) in enumerate(V)]
+        return [lambdadot, xdots...]
+    end
+end
+
+@doc raw"""
+    function embed(M::Segre{𝔽, V}, v)
+
+Embed ``p \doteq (\lambda, x_1, \dots, x_d)`` in ``𝔽^{n_1 \times \dots \times n_d}`` using the Krönecker product:
+````math
+    (\lambda, x_1, \dots, x_d) \mapsto \lambda x_1 \otimes \dots \otimes x_d.
+````
+"""
+function embed(M::Segre{𝔽,V}, p) where {𝔽,V}
+    return kronecker(p...)[:]
+end
+
+@doc raw"""
+    function embed_vector(M::Segre{𝔽, V}, p, v)
+
+Embed tangent vector ``v = (\nu, u_1, \dots, u_d)`` at ``p \doteq (\lambda, x_1, \dots, x_d)`` in ``𝔽^{n_1 \times \dots \times n_d}`` using the Krönecker product:
+````math
+    (\nu, u_1, \dots, u_d) \mapsto \nu x_1 \otimes \dots \otimes x_d + \lambda u_1 \otimes x_2 \otimes \dots \otimes x_d + \dots + \lambda x_1 \otimes \dots \otimes x_{d - 1} \otimes u_d.
+````
+"""
+function embed_vector(M::Segre{𝔽,V}, p, v) where {𝔽,V}
+
+    # Product rule
+    return sum([
+        kronecker([i == j ? xdot : x for (j, (x, xdot)) in enumerate(zip(p, v))]...)[:] for
+        (i, _) in enumerate(p)
+    ])
+end
+
+@doc raw"""
+    function m(M::Segre{ℝ, V}, p, q)
+
+Let ``p \doteq (\lambda, x_1, \dots, x_d)``, ``q \doteq (\mu, y_1, \dots, y_d) \in \mathcal{S}``.
+Then
+````math
+    m(p, q) = \sqrt{\sphericalangle(x_1, y_1)^2 + \dots + \sphericalangle(x_d, y_d)^2}
+````
+is the distance between ``(x_1, \dots, x_d)`` and ``(y_1, \dots, y_d)`` on ``S^{n_1 - 1} \times \dots \times S^{n_d - 1}``.
+"""
+function m(M::Segre{ℝ,V}, p, q) where {V}
```

Comment:
Should this function maybe be called `distance(M, p, q)`? Or am I misreading the doc string? What is the function used for? We should otherwise maybe find a better name.

Reply:
If it is only used in the function below, then it should maybe also be only part of that function and be defined and explained in the doc string there.

Reply:
`m()` is an unhelpfully named helper function that is used in `distance()`, `log()`, `connected_by_geodesic()`, and `closest_representation()`. I can come up with a better name for it maybe, but it's really just a shorthand for a specific expression that pops up alot in the formulas.

Reply:
I think will call it `spherical_angle_sum()` and explain in its docstring that it's a helper function.

Reply:
You do not have to explicitly state that it is a helper function.
The name sounds good, add the formula as well, then you could `@ref` to the function in all 4 mentioned functions above :)
And sure if it is used 4 times it is worth its own functions.